rfcs/text/0030/process.yml

```diff
@@ -0,0 +1,158 @@
+# license agreements. See the NOTICE file distributed with
+# this work for additional information regarding copyright
+# ownership. Elasticsearch B.V. licenses this file to you under
+# the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# 	http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+---
+- name: process
+  reusable:
+    top_level: true
+    expected:
+      - at: process
+        as: entry_leader
+        short_override: Information about the entry session leader process.
+      - at: process
+        as: session_leader
+        short_override: Information about the session leader process. In some cases this may be the same information as entry_leader.
+      - at: process
+        as: group_leader
+        short_override: Information about the process group leader process. In some cases this may be the same as the top level process.
+      - at: process.entry_leader
+        as: parent 
+        short_override: Information about the entry leaders parent. Only pid, start and entity_id fields are set.
+      - at: process.parent
+        as: group_leader
+        short_override: Information about the parent's process group leader process. Only pid, start and entity_id fields are set.
+
+  fields:
+    # Deprecation note: Because we are nesting process.group_leader above, it makes sense to deprecate process.pgid as this information can be accessed via process.group_leader.pid
+    # - name: pgid  #to be removed
+
+    - name: previous_args
+      level: extended
+      type: keyword
+      short: Array of process arguments the previous fork or exec.
+      description: >
+        Array of process arguments, starting with the absolute path to the executable.
+
+        May be filtered to protect sensitive information.
+      example: "[\"/usr/bin/ssh\", \"-l\", \"user\", \"10.0.0.16\"]"
+      normalize:
+        - array
+
+    - name: previous_args_count
+      level: extended
+      type: long
+      short: Length of the process.previous_args array.
+      description: >
+        Length of the process.previous_args array.
+
+        This field can be useful for querying or performing bucket analysis on
+        how many arguments were provided to start a process.
+        More arguments may be an indication of suspicious activity.
+      example: 4
+
+    - name: previous_executable
+      level: extended
+      type: keyword
+      description: >
+        Absolute path to the process previous executable.
+      example: /usr/bin/ssh
+      multi_fields:
+        - type: match_only_text
+          name: text
```

Comment:
Am I right to assume these two entries (`previous_args_count` and `previous_executable`) are meant to represent information about a process from before an exec call? Are the other "before the exec" members?

Rather than calling out these two fields, why not instead start a new `previous` member of the `process` fieldset which contains an array of process fieldsets? That way other `process` fieldset members can easily be included as well. From something like invoking `/bin/env python` from a bash shell the JSON would look like:

```json
{
  "process": {
    "executable": "/usr/bin/python",
    "args_count": 1,
    "previous": [
      { "executable": "/bin/env", "args_count": 2},
      { "executable": "/bin/bash", "args_count": 1},
    ]
  }
```

Reply:
I like this idea. Thoughts? @m-sample

Reply:
Unfortunately expressing an array of fieldsets is not possible with current ECS tooling. cc @kgeller brought this up in the comment above (https://github.com/elastic/ecs/pull/1684#issuecomment-1013209040), since we decided to use an array of objects for the process.fds array. I guess it's worth considering whether the ECS team wants to include support for this.

Reply:
@mitodrummer I think it's worth discussing as a group (ECS and cmd) about the re-use of a fieldset as an array, if it's something you want to pursue for this work.

Reply:
I love this idea. I think the limited search is ok (ie no need for nested/doc per entry in previous).   I would suggest we also put a timestamp in there for when those execs took place.    

Would anything be implied by the array entry ordering (time?) Doesn't seem like a good idea esp. if we have timestamps in there.

This pattern of previous tracking values is also relevant to group_leader (can change) and even parent (reparented if orig parent dies).  Even session_leader can change (once only when call setsid, and from prev session to *this* process being session leader)

Reply:
Does everyone agrees on using an array at `process.previous`?

If so, I believe these are the next steps to capture that decision here in the proposed field defs:

* Remove the `process.previous_args`,`process.previous_args_count`, `process.previous_executable` fields.
* Add another `process.*` reuse:

```
      - at: process
        as: previous
        short_override: An array of previous processes replaced via exec...
```

We can revisit syntax later (e.g., the stage 2 PR) once the ECS team decides how to state fieldset reuse may be an array.

Reply:
Ya, I think we are aligned here. Will push an update to add this. Thanks @ebeahan

Reply:
I guess I should do the same for process.file_descriptions? Currently it's in process.yml as array of type object, but i can move back to file_description.yaml reusable section. cc @ebeahan